src/flow.ts

```diff
@@ -606,6 +614,65 @@ export class Flow {
     localFlags[index] = flags & ~flag;
   }
 
+  /** set type assign */
+  setAssignType(expr: ExpressionRef, element: TypedElement, type: Type): void {
+    if (type && !type.isReference) return;
+    this.setNarrowedType(element, type);
+    this.conditionalNarrowedType.setAssignType(expr, element, type);
+  }
+  /** set type narrow */
+  setNarrowedType(element: TypedElement, type: Type): void {
+    if (!type.isReference) return;
+    if (this.narrowedTypes == null) {
+      this.narrowedTypes = new NarrowedTypeMap();
+    }
```

Comment:
I guess we could use this now:
```suggestion
    this.narrowedTypes = this.narrowedTypes || new NarrowedTypeMap();
```
Which will simplify further when we will support `||=` and `&&=`